dandiapi/api/models/asset.py

```diff
@@ -104,30 +110,60 @@ class Status(models.TextChoices):
     )
     published = models.BooleanField(default=False)
 
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                name='exactly-one-blob',
+                check=Q(blob__isnull=True, zarr__isnull=False)
+                | Q(blob__isnull=False, zarr__isnull=True),
+            )
+        ]
+
+    @property
+    def is_blob(self):
+        return self.blob is not None and self.zarr is None
+
+    @property
+    def is_zarr(self):
+        return self.zarr is not None and self.blob is None
+
     @property
     def size(self):
-        return self.blob.size
+        if self.is_blob:
+            return self.blob.size
+        else:
+            return self.zarr.size
```

Comment:
and there is no benefit (or possibility?) to properly decouple via OOP, i.e. to have `BlobAsset` and `ZarrAsset` with those subclasses defining proper behavior instead of resorting to such `if/else`s?

Reply:
it's possible in Django (https://docs.djangoproject.com/en/3.2/ref/contrib/contenttypes/), but it's considerably more obnoxious to use. Since the cost here is basically just a handful of icky if statements, I decided to use the simple way for now. If we ever need a new blob type, will will probably need to use OOP.